proxy/proxy.go

```diff
@@ -365,3 +368,54 @@ func (proxy *Proxy) weaveCIDRsFromConfig(config *docker.Config, hostConfig *dock
 	}
 	return nil, nil
 }
+
+func (proxy *Proxy) addWeaveWaitVolume(hostConfig *docker.HostConfig) {
+	var binds []string
+	for _, bind := range hostConfig.Binds {
+		s := strings.Split(bind, ":")
+		if len(s) >= 2 && s[1] == "/w" {
+			continue
+		}
+		binds = append(binds, bind)
+	}
+	hostConfig.Binds = append(binds, fmt.Sprintf("%s:/w:ro", proxy.weaveWaitVolume))
+}
+
+func (proxy *Proxy) setWeaveDNS(hostConfig *docker.HostConfig, hostname, dnsDomain string) error {
+	hostConfig.DNS = append(hostConfig.DNS, proxy.dockerBridgeIP)
+
+	if len(hostConfig.DNSSearch) == 0 {
+		if hostname == "" {
+			hostConfig.DNSSearch = []string{dnsDomain}
+		} else {
+			hostConfig.DNSSearch = []string{"."}
+		}
+	}
+
+	return nil
+}
+
+func (proxy *Proxy) getDNSDomain() (domain string, running bool) {
+	domain = nameserver.DefaultDomain
+	running = proxy.WithDNS
+	weaveContainer, err := proxy.client.InspectContainer("weave")
+	if err != nil ||
+		weaveContainer.NetworkSettings == nil ||
+		weaveContainer.NetworkSettings.IPAddress == "" {
+		return
+	}
+
+	url := fmt.Sprintf("http://%s:%d/domain", weaveContainer.NetworkSettings.IPAddress, router.HTTPPort)
+	resp, err := http.Get(url)
+	if err != nil || resp.StatusCode != http.StatusOK {
+		return
+	}
+	defer resp.Body.Close()
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return
+	}
+
+	return string(b), true
+}
```

Comment:
Surely this function needs to pay attention to `proxy.With*out*DNS` too. Also, `running` is misleading as a return value name, since it will be set to true if `proxy.WithDNS==true` even when weave isn't running. And perhaps we should ditch that bool return value altogether and instead return "" to indicate that dns settings should be left alone.